README.md
Capistrano Taffy (Database, SSH recipes)
================================================

**Capistrano recipes for deploying databases and other common tasks (managing database.yml, importing/exporting/transfering databases, SSH authorization etc.)**

Features
------------------------------------------------

* Adds database transfer recipes (via [`Taps`]("http://github.com/ricardochimal/taps"))
* Authorize SSH access
* Manage `database.yml` (Soon.)

[`Taps`]("http://github.com/ricardochimal/taps") is great, but having to SSH into my deployment to run the `Taps` server, as well as 
figure out the proper local/remote database urls, is a pain. I knew the [`Heroku`]("http://github.com/heroku/heroku") gem
had it figured out; I present the Capistrano friendly version.

If you are new to `Taps`, check out this [introduction to `Taps`]("http://adam.blog.heroku.com/past/2009/2/11/taps_for_easy_database_transfers/") by [Adam Wiggins]("http://github.com/adamwiggins").

Installation
------------------------------------------------

    gem install cap-taffy

Database Transfer
------------------------------------------------

> _Dependency:_ The [`Taps`]("http://github.com/ricardochimal/taps") gem is required on any server(s) you'll be transferring databases to (`:app` role) including your development machine (where you'll be running `cap` tasks from). Run:

>     gem install taps

> _Dependency:_ The [`Heroku`]("http://github.com/heroku/heroku") gem is required on your development machine.

>     gem install heroku

### Usage

To start, add the following to your `Capfile`

    require 'cap-taffy/db'

`Taffy` will start a `Taps` server on port 5000 by default, so make sure port 5000 (or w/e port you end up using) is open on your `:app` server(s)

Then you can use:

    cap db:push         # Or to specify a different port: cap db:push -s taps_port=4321
    cap db:pull         # Or to specify a different port: cap db:pull -s taps_port=4321
    

#### SSH Local Forwarding

> Some of you may not be able/want to open up ports on your servers. Instead, you can run:

>     ssh -N -L[port]:127.0.0.1:[port] [user]@[remote-server]
> And then run:

>     cap db:push -s taps_port=[port] -s local=true
> Substituing the appropriate values for [port], [user], and [remote-server].
> #### Sample Usage
> >     ssh -N -L4321:127.0.0.1:4321 henry@load-test
> >     cap db:push -s taps_port=4321 -s local=true

SSH Access
------------------------------------------------

#### Usage

Add the following to your `Capfile`

    require 'cap-taffy/ssh'

Using a public key generated from `ssh-keygen` (e.g. `ssh-keygen -t rsa`), to authorize access:

    cap ssh:authorize       # authorizes local public key for SSH access to remote server(s)


Managing `database.yml`
------------------------------------------------

> Much needed and coming soon.

Credits
------------------------------------------------
Thanks to developers of the [`Taps`]("http://adam.blog.heroku.com/past/2009/2/11/taps_for_easy_database_transfers/") gem and the [`Heroku`]("http://github.com/heroku/heroku") gem. And of course, Capistrano, awesome!

History.txt
== 0.0.3 / 2010-01-26
* 1 major fix
  * Fixed for Heroku 1.6.3

== 0.0.2 / 2009-09-14
* 1 minor enhancement
  * Added authorizing SSH public keys on remote server(s).

== 0.0.1 / 2009-09-14

* 1 major enhancement
  * Initial release.
  * Added db push/pull.


spec/spec_helper.rb
require File.expand_path(File.join(File.dirname(__FILE__), %w[.. lib cap-taffy]))

module Capistrano # :nodoc:
end

Spec::Runner.configure do |config|
  # == Mock Framework
  #
  # RSpec uses it's own mocking framework by default. If you prefer to
  # use mocha, flexmock or RR, uncomment the appropriate line:
  #
  config.mock_with :mocha
  # config.mock_with :flexmock
  # config.mock_with :rr
end

Capistrano.send(:remove_const, "Configuration") rescue nil
Capistrano.const_set("Configuration", Class.new)
Capistrano::Configuration.class_eval do
  def self.instance
    @instance ||= Capistrano::Configuration.new
  end

  def load(*args, &block)
    instance_eval(&block) if block_given?
  end

  def namespaces
    @namespaces ||= {}
  end

  def namespace(name, &block)
    namespaces[name].instance_eval(&block)
  end
end

class String # :nodoc:
  def demodulize
    gsub(/^.*::/, '')
  end
end

module CapistranoHelpers
  def self.included(base) # :nodoc:
    base.extend CapistranoHelpers::ClassMethods
  end

  # Stubs the Capistrano Logger and yields to the block
  def with_logger(&blk) # :yields:
    logger_class = Class.new
    logger = mock()
    logger.stub_everything
    logger_class.stubs(:new).returns(logger)
    Capistrano.const_set("Logger", logger_class)
    yield
  ensure
    Capistrano.send(:remove_const, "Logger") rescue nil
  end

  # Helper for common operations in db tasks
  def db_with_expected_options(options) # :nodoc:
    @namespace.stubs(:fetch).with(:user).returns(options[:login])
    @namespace.instance_variable_set(:@remote_database_url, options[:remote_database_url])
    @namespace.instance_variable_set(:@local_database_url, options[:local_database_url])
  end

  # Stubs the variables hash used by Capistrano to accept command-line parameters
  def namespace_with_variables(variables) # :nodoc:
    @namespace.stubs(:variables).returns(variables)
  end

  # Creates an expectation for the Capistrano namespace/task <tt>instance</tt> for the <tt>:run</tt> action with <tt>*args</tt>.
  def run_with(instance, *args)
    instance.expects(:run).with(*args)
  end

  # The Capistrano <tt>:run</tt> action loops with <tt>channel</tt>, <tt>stream</tt>, and <tt>data</tt> until the channel is closed.
  # 
  # Passing in a <tt>:run</tt> expectation, modifies the expectation such that each subsequent invocation ("loop") yields <tt>channel</tt>, <tt>stream</tt>, and <tt>data</tt>
  #
  # ==== Parameters
  #
  # * <tt>:channel</tt> - A hash containing <tt>:host</tt>.
  # * <tt>:stream</tt> - A stream object.
  # * <tt>:data</tt> - A data object, usually a String.
  def simulating_run_loop_with(options={}, &blk) # :yields:
    channel = options[:channel] || {:host => "192.168.1.20"}
    stream = options[:stream]
    data = options[:data]

    blk.call.then.yields(channel, stream, data)
    [channel, stream, data]
  end

  module ClassMethods
    # Used in specs to test Capistrano tasks.
    #
    # 
    # Code defined in the task will be executed automatically on load. (Note the <tt>yields</tt>, see Mocha#yields[http://mocha.rubyforge.org/classes/Mocha/Expectation.html#M000043])
    #
    # ==== Parameters
    #
    # * <tt>:it</tt> - The description for the current example group.
    # * <tt>:in</tt> - Specifies the module under test (as well as the namespace the task is defined in). The namespace will be deduced with the downcase of <tt>:in</tt>
    #
    # ==== Examples
    #
    #   for_task :detect, :roles => :app, :in => :Somewhere, :it => "should so something" do
    #     @namespace # refers to the current task under test (in Capistrano tasks are executed on Capistrano::Namespaces::Namespace instances)
    #     @mod       # refers to module CapTaffy::Somewhere
    #   end
    def for_task(task_name, options, &block)
      description = options.delete(:it)
      namespace = options.delete(:in)

      context ":#{task_name.to_s} task" do
        before do
          @namespace = Capistrano::Configuration.instance.namespaces[namespace.to_s.downcase.to_sym] = mock()
          @namespace.stubs(:desc)
          @namespace.stubs(:task)
          args = [task_name, options]
          unless options.empty?
            @namespace.expects(:task).with(*args).yields
          else
            @namespace.expects(:task).with(task_name).yields
          end

          @mod = Module.new # The module under test
          CapTaffy.const_set(namespace, @mod)
        end

        it description, &block

        after do
          const_name = @mod.to_s.demodulize
          CapTaffy.send(:remove_const, const_name)
        end
      end
    end
  end
end

module TaffyHelpers
  def self.included(base) # :nodoc:
    base.extend TaffyHelpers::ClassMethods
  end

  # A simple helper for mocking a quick object
  # 
  # Usage:
  #   taps_client_who(:expects, :do_something)
  def taps_client_who(method_symbol, *args)
    client = mock()
    client.send(method_symbol, *args)
    client
  end

  module ClassMethods
    # A wrapper for running CapTaffy::Db::run
    def running_db_it(message, &blk)
      context "when running db" do
        before do
          @capistrano = mock()
        end

        # See CapistranoHelpers
        def capistrano_run_with(*args)
          run_with(@capistrano, *args) 
        end

        it message, &blk

        after do

        end
      end
    end
  end
end


spec/cap-taffy_spec.rb
require File.join(File.dirname(__FILE__), %w[spec_helper])

describe CapTaffy do
  it "should get the version" do
    CapTaffy::VERSION.should == CapTaffy.version
  end
end


spec/cap-taffy/db_spec.rb
require File.join(File.dirname(__FILE__), %w[.. spec_helper])

module CapTaffy
  describe 'Db' do
    context "when loaded" do
      include CapistranoHelpers
      include TaffyHelpers

      before do
        CapTaffy.send(:remove_const, "Db") rescue nil
      end

      it "should load in capistrano configuration instance" do;
        Capistrano::Configuration.instance.expects(:load)

        load 'lib/cap-taffy/db.rb'
      end

      it "should define :db namespace" do
        Capistrano::Configuration.instance.expects(:namespace).with(:db)

        load 'lib/cap-taffy/db.rb'
      end

      for_task :detect, :roles => :app, :in => :Db, :it => "should be defined" do
        @mod.expects(:remote_database_url).returns("remote_db_url")
        @mod.expects(:local_database_url).returns("local_db_url")
        
        load 'lib/cap-taffy/db.rb'
        
        @namespace.instance_variable_get(:@remote_database_url).should == "remote_db_url"
        @namespace.instance_variable_get(:@local_database_url).should == "local_db_url"
      end

      for_task :detect, :roles => :app, :in => :Db, :it => "should be safe if dry run" do
        @mod.expects(:remote_database_url).raises Exception
        @mod.expects(:local_database_url).raises Exception
        @namespace.stubs(:dry_run).returns(true)

        load 'lib/cap-taffy/db.rb'

        @namespace.instance_variable_get(:@remote_database_url).should == ""
        @namespace.instance_variable_get(:@local_database_url).should == ""
      end

      for_task :detect, :roles => :app, :in => :Db, :it => "should raise as usual if not dry run" do
        class SpecialError < StandardError; end
        @mod.expects(:remote_database_url).raises SpecialError
        @namespace.stubs(:dry_run).returns(false)

        lambda { load 'lib/cap-taffy/db.rb' }.should raise_error(SpecialError)
      end

      def load_taffy_db # :nodoc:
        with_logger do
          load 'lib/cap-taffy/db.rb'
        end
      end

      for_task :push, :roles => :app, :in => :Db, :it => "should send taps client cmd_send" do
        options = {:remote_database_url => "remote", :local_database_url => "local", :port => nil, :login => "a_user", :password => "a_pass"}
        @namespace.expects(:detect)
        namespace_with_variables(:taps_port => nil)
        db_with_expected_options(options)
        @mod.expects(:tmp_pass).with(@namespace.fetch(:user)).returns(options[:password])
        @mod.expects(:run).with(@namespace, options).yields(taps_client_who(:expects, :cmd_send))

        load_taffy_db
      end

      for_task :push, :roles => :app, :in => :Db, :it => "should use cli argument for port" do
        options = {:remote_database_url => "remote", :local_database_url => "local", :port => 1234, :login => "a_user", :password => "a_pass"}
        @namespace.expects(:detect)
        namespace_with_variables(:taps_port => 1234)
        db_with_expected_options(options)
        @mod.expects(:tmp_pass).with(@namespace.fetch(:user)).returns(options[:password])
        @mod.expects(:run).with(@namespace, options)

        load_taffy_db
      end

      for_task :push, :roles => :app, :in => :Db, :it => "should force 127.0.0.1 (local) for ssh local forwarding" do
        options = {:remote_database_url => "remote", :local_database_url => "local", :port => 1234, :login => "a_user", :password => "a_pass"}
        @namespace.expects(:detect)
        namespace_with_variables(:taps_port => 1234, :local => true)
        db_with_expected_options(options)
        @mod.expects(:tmp_pass).with(@namespace.fetch(:user)).returns(options[:password])
        @mod.expects(:run).with(@namespace, options.merge(:local => true))

        load_taffy_db
      end
      
      for_task :pull, :roles => :app, :in => :Db, :it => "should send taps client cmd_receive" do
        options = {:remote_database_url => "remote", :local_database_url => "local", :port => nil, :login => "a_user", :password => "a_pass"}
        @namespace.expects(:detect)
        namespace_with_variables(:taps_port => nil)
        db_with_expected_options(options)
        @mod.expects(:tmp_pass).with(@namespace.fetch(:user)).returns(options[:password])
        @mod.expects(:run).with(@namespace, options).yields(taps_client_who(:expects, :cmd_receive))

        load_taffy_db
      end
    end

    context "after capistrano loaded" do
      include CapistranoHelpers
      include TaffyHelpers

      before do
        Capistrano::Configuration.instance.expects(:namespace).with(:db)
        CapTaffy.send(:remove_const, "Db") rescue nil
        load 'lib/cap-taffy/db.rb'

        @conf = {"test"=>
          {"reconnect"=>false,
          "encoding"=>"utf8",
          "username"=>"root",
          "adapter"=>"postgresql",
          "database"=>"test_test",
          "pool"=>5,
          "password"=>"root",
          "host"=>"localhost"}}

        @options = {:remote_database_url => "remote", :local_database_url => "local", :port => nil, :login => "a_user", :password => "a_pass"}
      end

      it "should detect local database url" do
        File.expects(:exists?).returns(true)
        File.expects(:read).returns(mock())
        YAML.expects(:load).returns(@conf)
        env = 'test'
        Parse.expects(:database_url).with(@conf, env)

        Db.local_database_url(env)
      end

      it "should detect remote database url" do
        instance = mock()
        instance.stubs(:current_path).returns("/home/user/public_html/current")
        instance.expects(:capture).with("cat /home/user/public_html/current/config/database.yml")
        YAML.expects(:load).returns(@conf)
        env = 'test'
        Parse.expects(:database_url).with(@conf, env)

        Db.remote_database_url(instance, env)
      end

      it "should create temporary password from time and user" do
        Time.stubs(:now).returns("asdfasdfasdf")
        user = "me"
        tmp_pass = Digest::SHA1.hexdigest("--#{Time.now.to_s}--#{user}--")

        Db.tmp_pass(user).should == tmp_pass
      end

      it "should expose a basic taps client" do
        client = mock()
        local_database_url = mock()
        remote_url = mock()
        Taps::Config.expects(:chunksize=).with(1000)
        Taps::Config.expects(:database_url=).with(local_database_url)
        Taps::Config.expects(:remote_url=).with(remote_url)
        Taps::Config.expects(:verify_database_url)
        client.expects(:do_something)

        Taps::ClientSession.expects(:quickstart).yields(client)

        Db.taps_client(local_database_url, remote_url) do |client|
          client.do_something
        end
      end

      it "should define default port" do
        Db.default_server_port.should == 5000
      end

      it "should build server command" do
        remote_database_url, login, password, port = @options[:remote_database_url], @options[:login], @options[:password], @options[:port]

        Db.server_command(@options).should == "taps server #{remote_database_url} #{login} #{password}"
      end
      
      it "should build server command with port" do
        @options[:port] = 1234
        remote_database_url, login, password, port = @options[:remote_database_url], @options[:login], @options[:password], @options[:port]

        Db.server_command(@options).should == "taps server #{remote_database_url} #{login} #{password} --port=#{port}"
      end
      
      it "should build server command without port if same as default port" do
        @options[:port] = 5000
        remote_database_url, login, password = @options[:remote_database_url], @options[:login], @options[:password]

        Db.server_command(@options).should == "taps server #{remote_database_url} #{login} #{password}"
      end

      def parser_expects_uri_hash_to_url_with(login, password, host)
        parser = mock()
        Parse.expects(:new).at_least_once.returns(parser)
        parser.expects(:uri_hash_to_url).
          with('username' => login, 'password' => password, 'host' => host, 'scheme' => 'http', 'path' => '')
      end

      it "should build remote url (with some help)" do
        @options[:host] = "127.0.0.1"        
        parser_expects_uri_hash_to_url_with(@options[:login], @options[:password], "#{@options[:host]}:#{Db.default_server_port}").returns("remote_url/")

        Db.remote_url(@options)
      end

      it "should build remote url with different port" do
        @options[:host] = "127.0.0.1"
        @options[:port] = 1234
        parser_expects_uri_hash_to_url_with(@options[:login], @options[:password], "#{@options[:host]}:#{@options[:port]}").returns("remote_url")

        Db.remote_url(@options)
      end

      it "should remove trailing slash from remote url" do
        @options[:host] = "127.0.0.1"
        parser_expects_uri_hash_to_url_with(@options[:login], @options[:password], "#{@options[:host]}:#{Db.default_server_port}").returns("remote_url/")

        Db.remote_url(@options).should == "remote_url"
      end

      running_db_it "should run with capistrano" do
        capistrano_run_with(Db.server_command(@options))

        Db.run(@capistrano, @options)
      end

      running_db_it "should do something to taps client" do
        channel, stream, data = simulating_run_loop_with :data => ">> Listening on 0.0.0.0:5000, CTRL+C to stop\r\n" do
          capistrano_run_with(Db.server_command(@options))
        end
        channel.expects(:close)

        Db.expects(:remote_url).with(@options.merge(:host => channel[:host], :port => 5000)).returns("remote_url")
        Db.expects(:taps_client).with("local", "remote_url").yields(taps_client_who(:expects, :do_something))
        Db.run(@capistrano, @options) do |client|
          client.do_something
        end

        channel[:status].should == 0
      end

      running_db_it "should run taffy on different port" do
        @options[:port] = 1234

        channel, stream, data = simulating_run_loop_with :data => ">> Listening on 0.0.0.0:1234, CTRL+C to stop\r\n" do
          capistrano_run_with(Db.server_command(@options))
        end
        channel.expects(:close)
        Db.expects(:remote_url).with(@options.merge(:host => channel[:host])).returns("remote_url")
        Db.expects(:taps_client).with("local", "remote_url").yields(taps_client_who(:expects, :do_something))

        Db.run(@capistrano, @options) do |client|
          client.do_something
        end

        channel[:status].should == 0
      end

      running_db_it "should not do anything until taps sinatra server is running" do
        simulating_run_loop_with :data => "asdfasdf" do
          capistrano_run_with(Db.server_command(@options))
        end

        client = mock()
        client.expects(:do_something).never

        Db.run(@capistrano, @options) do |client|
          client.do_something
        end

        channel, stream, data = simulating_run_loop_with :data => ">> Listening on 0.0.0.0:5000, CTRL+C to stop\r\n" do
          capistrano_run_with(Db.server_command(@options))
        end
        channel.expects(:close)
        Db.expects(:remote_url).with(@options.merge(:host => channel[:host], :port => 5000)).returns("remote_url")
        Db.expects(:taps_client).with("local", "remote_url").yields(taps_client_who(:expects, :do_something))

        Db.run(@capistrano, @options) do |client|
          client.do_something
        end

        channel[:status].should == 0
      end

      running_db_it "should force 127.0.0.1 (local) for remote url" do
        channel, stream, data = simulating_run_loop_with :data => ">> Listening on 0.0.0.0:5000, CTRL+C to stop\r\n" do
          capistrano_run_with(Db.server_command(@options))
        end
        channel.expects(:close)

        Db.expects(:remote_url).with(@options.merge(:host => '127.0.0.1', :port => 5000)).returns("remote_url")
        Db.expects(:taps_client).with("local", "remote_url").yields(taps_client_who(:expects, :do_something))
        Db.run(@capistrano, @options.merge(:local => true)) do |client|
          client.do_something
        end

        channel[:status].should == 0
      end
    end
  end
end


lib/cap-taffy/ssh.rb
require File.join(File.dirname(__FILE__), %w[.. cap-taffy]) unless defined?(CapTaffy)

module CapTaffy::SSH

end

Capistrano::Configuration.instance.load do
  namespace :ssh do
    desc <<-DESC
      Authorize SSH access for local computer on remote computers(s).
    DESC
    task :authorize do
      public_key = File.read(File.expand_path(File.join(%w[~/ .ssh id_rsa.pub]))).chop

      run %Q[if [ ! -f ~/.ssh/authorized_keys ]; then mkdir -p ~/.ssh && touch ~/.ssh/authorized_keys && chmod 600 ~/.ssh/authorized_keys; fi && if [ -z "$(grep "^#{public_key}$" ~/.ssh/authorized_keys)" ]; then echo "#{public_key}" >> ~/.ssh/authorized_keys && echo "Public key on '$CAPISTRANO:HOST$' authorized at '#{Time.now.to_s}'"; else echo "Public key on '$CAPISTRANO:HOST$' is already authorized."; fi]
    end
  end
end


lib/cap-taffy/parse.rb
require 'rubygems'
begin
  require 'heroku'
  require 'heroku/command'
  require 'heroku/commands/base'
  require 'heroku/commands/db'
rescue LoadError
  error "Install the Heroku gem. On most systems this will be:\nsudo gem install heroku"
end

module CapTaffy
  class Parse < Heroku::Command::Db
    class << self
      attr_accessor :instance

      # Modified from :parse_database_yml in heroku/command/db.rb
      #
      # Accepts a complete +db_config+ hash and +env+ and parses for a database_url accordingly.
      def database_url(db_config, env)
        raise Invalid, "please pass me a valid Hash loaded from a database YAML file" unless db_config
        conf = db_config[env]
        raise Invalid, "missing '#{env}' database in #{db_config.inspect}" unless conf

        self.instance ||= CapTaffy::Parse.new

        case conf['adapter']
        when 'sqlite3'
          return "sqlite://#{conf['database']}"
        when 'postgresql'
          uri_hash = self.instance.conf_to_uri_hash(conf)
          uri_hash['scheme'] = 'postgres'
          return self.instance.uri_hash_to_url(uri_hash)
        else
          return self.instance.uri_hash_to_url(self.instance.conf_to_uri_hash(conf))
        end
      end
    end

    # Override to do nothing on #new
    def initialize # :nodoc:

    end

    # Override to pass-through on #escape
    def escape(string)
      string
    end

    public :uri_hash_to_url, :conf_to_uri_hash

    class Invalid < RuntimeError # :nodoc:
    end
  end
end


lib/cap-taffy/db.rb
require 'rubygems'
begin
  gem 'taps', '>= 0.2.8', '< 0.3.0'
  require 'taps/client_session'
rescue LoadError
  error "Install the Taps gem to use db commands. On most systems this will be:\nsudo gem install taps"
end

require File.join(File.dirname(__FILE__), %w[.. cap-taffy]) unless defined?(CapTaffy)
require File.join(File.dirname(__FILE__), 'parse')
require 'digest/sha1'

module CapTaffy::Db
  extend self

  # Detects the local database url for +env+.
  #
  # Looks for <tt>config/database.yml</tt>.
  def local_database_url(env)
    return "" unless File.exists?(Dir.pwd + '/config/database.yml')
    db_config = YAML.load(File.read(Dir.pwd + '/config/database.yml'))

    CapTaffy::Parse.database_url(db_config, env)
  end

  # Detects the remote database url for +env+ and the current Capistrano +instance+.
  #
  # Looks for <tt>config/database.yml</tt> in the +current_path+.
  def remote_database_url(instance, env)
    db_yml = instance.capture "cat #{instance.current_path}/config/database.yml"
    db_config = YAML::load(db_yml)

    CapTaffy::Parse.database_url(db_config, env)
  end

  # The default server port the Taps server is started on.
  def default_server_port
    5000
  end

  # Generates the remote url used by Taps push/pull.
  #
  # ==== Parameters
  # 
  # * <tt>:login, :password, :host, :port</tt> - See #run.
  #
  # ==== Examples
  #
  #   login = fetch(:user)
  #   password = tmp_pass(login)                                               # returns asdkf239udjhdaks (for example)
  #   remote_url(:login => login, :password => password, :host => 'load-test') # returns http://henry:asdkf239udjhdaks@load-test:5000
  def remote_url(options={})
    host = options[:host]
    port = options[:port] || default_server_port
    host += ":#{port}"
    url = CapTaffy::Parse.new.uri_hash_to_url('username' => options[:login], 'password' => options[:password], 'host' => host, 'scheme' => 'http', 'path' => '')

    url.sub(/\/$/, '')
  end

  # Generates a temporary password to be used for the Taps server command.
  def tmp_pass(user)
    Digest::SHA1.hexdigest("--#{Time.now.to_s}--#{user}--")
  end

  # A quick start for a Taps client.
  #
  # <tt>local_database_url</tt> and <tt>remote_url</tt> refer to the options for the Taps gem (see #run).
  def taps_client(local_database_url, remote_url, &blk) # :yields: client
    Taps::Config.chunksize = 1000
    Taps::Config.database_url = local_database_url
    Taps::Config.remote_url = remote_url
    Taps::Config.verify_database_url

    Taps::ClientSession.quickstart do |client|
      yield client
    end
  end

  # Generates the server command used to start a Taps server
  #
  # ==== Parameters
  # * <tt>:remote_database_url, :login, :password</tt> - See #run.
  # * <tt>:port</tt> - The +port+ the Taps server is on. If given and different from #default_server_port, appends <tt>--port=[port]</tt> to command.
  def server_command(options={})
    remote_database_url, login, password, port = options[:remote_database_url], options[:login], options[:password], options[:port]
    port_argument = ''
    port_argument = " --port=#{port}" if port && port != default_server_port

    "taps server #{remote_database_url} #{login} #{password}#{port_argument}"
  end

  # The meat of the operation. Runs operations after setting up the Taps server.
  # 
  # 1. Runs the <tt>taps</tt> taps command to start the Taps server (assuming Sinatra is running on Thin)
  # 2. Wait until the server is ready 
  # 3. Execute block on Taps client
  # 4. Close the connection(s) and bid farewell.
  #
  # ==== Parameters
  # * <tt>:remote_database_url</tt> - Refers to local database url in the options for the Taps server command (see Taps Options).
  # * <tt>:login</tt> - The login for +host+. Usually what's in <tt>set :user, "the user"</tt> in <tt>deploy.rb</tt>
  # * <tt>:password</tt> - The temporary password for the Taps server.
  # * <tt>:port</tt> - The +port+ the Taps server is on. If not given, defaults to #default_server_port.
  # * <tt>:local_database_url</tt> - Refers to the local database url in the options for Taps client commands (see Taps Options).
  #
  # ==== Taps Options
  #
  # <tt>taps</tt>
  #   server <local_database_url> <login> <password> [--port=N]        Start a taps database import/export server
  #   pull <local_database_url> <remote_url> [--chunksize=N]           Pull a database from a taps server
  #   push <local_database_url> <remote_url> [--chunksize=N]           Push a database to a taps server
  #
  # ==== Examples
  #
  #   task :push do
  #     login = fetch(:user)
  #     password = Time.now.to_s
  #     CapTaffy.Db.run(self, { :login => login, :password => password, :remote_database_url => "sqlite://test_production", :local_database_url => "sqlite://test_development" }) do |client|
  #       client.cmd_send
  #     end
  #   end
  def run(instance, options = {} , &blk) # :yields: client
    options[:port] ||= default_server_port
    remote_database_url, login, password, port, local_database_url = options[:remote_database_url], options[:login], options[:password], options[:port], options[:local_database_url]
    force_local = options.delete(:local)

    data_so_far = ""
    instance.run CapTaffy::Db.server_command(options) do |channel, stream, data|
      data_so_far << data
      if data_so_far.include? ">> Listening on 0.0.0.0:#{port}, CTRL+C to stop"
        host = force_local ? '127.0.0.1' : channel[:host]
        remote_url = CapTaffy::Db.remote_url(options.merge(:host => host))

        CapTaffy::Db.taps_client(local_database_url, remote_url) do |client|
          yield client
        end

        data_so_far = ""
        channel.close
        channel[:status] = 0
      end
    end
  end

  class InvalidURL < RuntimeError # :nodoc:
  end
end

Capistrano::Configuration.instance.load do
  namespace :db do
    # Executes given block.
    # If this is a dry run, any raised exceptions will be caught and +returning+ is returned.
    # If this is not a dry run, any exceptions will be raised as expected.
    def dry_run_safe(returning = nil, &block) # :yields:
      begin
        yield
      rescue Exception => e
        raise e unless dry_run
        return returning
      end
    end

    task :detect, :roles => :app do
      @remote_database_url = dry_run_safe('') { CapTaffy::Db.remote_database_url(self, 'production') }
      @local_database_url = dry_run_safe('') { CapTaffy::Db.local_database_url('development') }
    end

    desc <<-DESC
      Push a local database into the app's remote database.
      
      Performs push from local development database to remote production database.
      Opens a Taps server on port 5000. (Ensure port is opened on the remote server).
      
        # alternately, specify a different port
        cap db:push -s taps_port=4321
        
      For the security conscious:
      
        # use ssh local forwarding (ensure [port] is available on both endpoints)
        ssh -N -L[port]:127.0.0.1:[port] [user]@[remote-server]
        
        # then push locally
        cap db:push -s taps_port=[port] -s local=true
    DESC
    task :push, :roles => :app do      
      detect

      login = fetch(:user)
      password = CapTaffy::Db.tmp_pass(login)

      logger = Capistrano::Logger.new
      logger.important "Auto-detected remote database: #{@remote_database_url}" if @remote_database_url != ''
      logger.important "Auto-detected local database: #{@local_database_url}" if @local_database_url != ''

      options = {:remote_database_url => @remote_database_url, :login => login, :password => password, :local_database_url => @local_database_url, :port => variables[:taps_port]}
      options.merge!(:local => true) if variables[:local]

      CapTaffy::Db.run(self, options) do |client|
        client.cmd_send
      end
    end

    desc <<-DESC
      Pull the app's database into a local database.
      
      Performs pull from remote production database to local development database.
      Opens a Taps server on port 5000. (Ensure port is opened on the remote server).
        
        # alternately, specify a different port
        cap db:pull -s taps_port=4321
        
      For the security conscious:

        # use ssh local forwarding (ensure [port] is available on both endpoints)
        ssh -N -L[port]:127.0.0.1:[port] [user]@[remote-server]

        # then pull locally
        cap db:pull -s taps_port=[port] -s local=true
    DESC
    task :pull, :roles => :app do
      detect

      login = fetch(:user)
      password = CapTaffy::Db.tmp_pass(login)

      logger = Capistrano::Logger.new
      logger.important "Auto-detected remote database: #{@remote_database_url}" if @remote_database_url != ''
      logger.important "Auto-detected local database: #{@local_database_url}" if @local_database_url != ''

      options = {:remote_database_url => @remote_database_url, :login => login, :password => password, :local_database_url => @local_database_url, :port => variables[:taps_port]}
      options.merge!(:local => true) if variables[:local]

      CapTaffy::Db.run(self, options) do |client|
        client.cmd_receive
      end
    end
  end
end
